presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlattenedStreamReader.java

```diff
@@ -0,0 +1,383 @@
+    @Nullable
+    private BooleanInputStream presentStream;
+
+    private boolean rowGroupOpen;
+
+    public MapFlattenedStreamReader(StreamDescriptor streamDescriptor, DateTimeZone hiveStorageTimeZone, AggregatedMemoryContext systemMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "stream is null");
+        this.hiveStorageTimeZone = requireNonNull(hiveStorageTimeZone, "hiveStorageTimeZone is null");
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        this.keyOrcType = streamDescriptor.getNestedStreams().get(0).getStreamType();
+        this.baseValueStreamDescriptor = streamDescriptor.getNestedStreams().get(1);
+    }
+
+    @Override
+    public void prepareNextRead(int batchSize)
+    {
+        readOffset += nextBatchSize;
+        nextBatchSize = batchSize;
+    }
+
+    @Override
+    public Block readBlock(Type type)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        if (readOffset > 0) {
+            if (presentStream != null) {
+                // skip ahead the present bit reader, but count the set bits
+                // and use this as the skip size for the data reader
+                readOffset = presentStream.countBitsSet(readOffset);
+            }
+            if (readOffset > 0) {
+                for (int i = 0; i < valueStreamReaders.size(); i++) {
+                    int valueReadOffset = inMapStreams.get(i).countBitsSet(readOffset);
+                    valueStreamReaders.get(i).prepareNextRead(valueReadOffset);
+                }
+            }
+        }
+
+        // We will use the offsetVector as the buffer to read the length values from lengthStream,
+        // and the length values will be converted in-place to an offset vector.
+        boolean[][] inMapVectors = new boolean[inMapStreams.size()][nextBatchSize];
+        boolean[] nullVector = null;
+
+        if (presentStream == null) {
+            for (int i = 0; i < inMapStreams.size(); i++) {
+                inMapStreams.get(i).getSetBits(nextBatchSize, inMapVectors[i]);
+            }
+        }
+        else {
+            nullVector = new boolean[nextBatchSize];
+            int nullValues = presentStream.getUnsetBits(nextBatchSize, nullVector);
+            if (nullValues != nextBatchSize) {
+                for (int i = 0; i < inMapStreams.size(); i++) {
+                    inMapStreams.get(i).getSetBits(nextBatchSize, inMapVectors[i], nullVector);
+                }
+            }
+        }
+
+        MapType mapType = (MapType) type;
+        Type keyType = mapType.getKeyType();
+        Type valueType = mapType.getValueType();
+
+        Block[] valueBlocks = new Block[valueStreamReaders.size()];
+
+        for (int i = 0; i < valueStreamReaders.size(); i++) {
+            int mapsContainingValue = 0;
```

Comment:
I used an int so that I could pass this as the batch size to prepareNextRead.  This might be a micro-optimization, I could just use the original batch size.  Just wanted to make sure that was clear, let me know what you think